File: windows-desktop/multimodal_adapters.py
"""Consent-first specifications for Arthur's future multimodal adapters.

These are deliberately declarative. They do not enumerate devices, request a
Windows permission, open a microphone/camera/screen stream, or contact a
provider. The installed application must obtain fresh user approval before an
adapter becomes eligible for a separate runtime implementation.
"""

from dataclasses import asdict, dataclass


@dataclass(frozen=True)
class AdapterSpec:
    identifier: str
    label: str
    input_kind: str
    default_state: str
    transport: str
    activation_requirement: str
    credential_requirement: str


ADAPTERS = (
    AdapterSpec(
        "speech_stream",
        "Speech-to-text and text-to-speech pipeline",
        "microphone / speaker",
        "disabled",
        "closed",
        "Select a route, approve microphone or speech-output use, and approve the chosen local engine or provider connection.",
        "No key for a local engine; a developer-owned provider key is required only for an approved provider route.",
    ),
    AdapterSpec(
        "vision_matrix",
        "Local camera vision matrix",
        "camera frames",
        "disabled",
        "closed",
        "Unlock the Spatial Room, select a camera, approve a visible local camera session, then start a time-bounded session.",
        "No key for a local-only adapter; a developer-owned key is required only before an approved external vision provider receives selected data.",
    ),
    AdapterSpec(
        "screen_share",
        "Explicit screen or window share",
        "selected display or window",
        "disabled",
        "closed",
        "Choose the exact display or window in the operating-system picker for each session and review the sharing boundary.",
        "No key for local capture; a developer-owned key is required only before approved external analysis.",
    ),
    AdapterSpec(
        "coordinate_stream",
        "Local coordinate revision relay",
        "Arthur layout JSON",
        "disabled",
        "closed",
        "Approve a local listener port, trusted client list, session lifetime, and local firewall boundary before a separate relay starts.",
        "No API key for a local listener; an authenticated relay credential is required before any remote synchronization.",
    ),
)


def list_adapter_contracts() -> list[dict]:
    """Return JSON-ready disabled adapter descriptions for UI and tests."""
    return [asdict(adapter) for adapter in ADAPTERS]


def adapter_contract(identifier: str) -> dict | None:
    """Return one non-activating adapter contract when it is declared."""
    return next((asdict(adapter) for adapter in ADAPTERS if adapter.identifier == identifier), None)

File: client/src/lib/multimodalAdapters.ts
/**
 * Arthur's browser-facing multimodal contract. It is declarative only: the
 * preview has no getUserMedia call, screen-capture request, socket, broker, or
 * provider client. A later installed-app adapter must obtain its own consent.
 */
export type MultimodalAdapterId = "speech_stream" | "vision_matrix" | "screen_share" | "coordinate_stream" | "environment_hub";

export type MultimodalAdapterContract = {
  id: MultimodalAdapterId;
  label: string;
  input: string;
  defaultState: "disabled";
  transport: "closed";
  activationRequirement: string;
  credentials: string;
};

export const multimodalAdapterContracts: readonly MultimodalAdapterContract[] = [
  { id: "speech_stream", label: "Speech streaming", input: "Microphone and speaker", defaultState: "disabled", transport: "closed", activationRequirement: "Choose a speech route, then separately approve the device and engine or provider.", credentials: "No key for local engines; developer-managed key only for an approved provider." },
  { id: "vision_matrix", label: "Vision matrix", input: "Camera frames", defaultState: "disabled", transport: "closed", activationRequirement: "Unlock the Spatial Room and approve a visible, time-bounded local camera session.", credentials: "No key for local-only handling; developer-managed key only for an approved external vision provider." },
  { id: "screen_share", label: "Screen or window share", input: "One user-selected display or window", defaultState: "disabled", transport: "closed", activationRequirement: "Choose the exact display or window in the operating-system picker for each session.", credentials: "No key for local capture; developer-managed key only before approved external analysis." },
  { id: "coordinate_stream", label: "Coordinate revision relay", input: "Arthur layout JSON", defaultState: "disabled", transport: "closed", activationRequirement: "Approve a loopback port, named local client, session lifetime, and firewall boundary.", credentials: "No key for an approved loopback listener; authenticated relay credentials for any remote sync." },
  { id: "environment_hub", label: "Home Assistant or MQTT hub", input: "One named scene or topic", defaultState: "disabled", transport: "closed", activationRequirement: "Enter an endpoint, developer credential, and one reviewed scene or topic, then approve each control action.", credentials: "Home Assistant long-lived access token or MQTT credentials only after explicit integration approval." },
] as const;

export function multimodalContractById(id: string): MultimodalAdapterContract | undefined {
  return multimodalAdapterContracts.find((contract) => contract.id === id);
}
